### PR TITLE
Split tropical_cyclone module

### PR DESCRIPTION
**Pull request by ThomasRoosli**
Changes proposed in this PR:
- The very large trop_cyclone.py file is split in two
- One part is the Trop_Cyclone class
- The other part is the windfields calculation using the different Holland or other models, the interface between the two submodules is one single function.

This PR fixes no specific issue, but is was mentioned in the discussion of https://github.com/CLIMADA-project/climada_python/pull/846

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [ ] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/develop/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Review.html#reviewer-checklist) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/develop/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Testing.html
[linter]: https://climada-python.readthedocs.io/en/latest/guide/Guide_continuous_integration_GitHub_actions.html#static-code-analysis

### CODE REVIEW

File: climada/hazard/tc_tracks.py [hunk 1]
```diff
@@ -1383,7 +1383,7 @@ def write_hdf5(self, file_name, complevel=5):
         df_attrs = pd.DataFrame([t.attrs for t in data], index=ds_combined["storm"].to_series())
         ds_combined = xr.merge([ds_combined, df_attrs.to_xarray()])
 
-        encoding = {v: dict(zlib=True, complevel=complevel) for v in ds_combined.data_vars}
+        encoding = {v: dict(zlib=False, complevel=complevel) for v in ds_combined.data_vars}
```
Comment (peanutfun) [inline climada/hazard/tc_tracks.py:1386]: Great! However, this turns off compression for all variables and might drastically increase the file size. We want to apply zlib to numerical data, so something like the following might be appropriate:

```python
def zlib(var):
    if np.issubdtype(var.dtype, float) or np.issubdtype(var.dtype, int):
        return True
    return False

encoding = {v: dict(zlib=zlib(v), complevel=complevel) for v in ds_combined.data_vars}
```

Comment (ThomasRoosli) [reply climada/hazard/tc_tracks.py]: Good point. I added such a function.

I also made the (priviously private) function that connects the now split modules a public function, so no private function is imported over modules. Any other comments?

File: climada/hazard/tc_tracks.py [hunk 2]
```diff
@@ -1383,7 +1383,7 @@ def write_hdf5(self, file_name, complevel=5):
         df_attrs = pd.DataFrame([t.attrs for t in data], index=ds_combined["storm"].to_series())
         ds_combined = xr.merge([ds_combined, df_attrs.to_xarray()])
 
-        encoding = {v: dict(zlib=True, complevel=complevel) for v in ds_combined.data_vars}
+        encoding = {v: dict(zlib=_zlib_from_DataArray(ds_combined[v]), complevel=complevel) for v in ds_combined.data_vars}
```
Comment (peanutfun) [inline climada/hazard/tc_tracks.py:1386]: Line too long (also, rename the function)
```suggestion
        encoding = {
            v: dict(zlib=_zlib_from_dataarray(ds_combined[v]), complevel=complevel)
            for v in ds_combined.data_vars
        }
```

File: climada/hazard/tc_tracks.py [hunk 3]
```diff
@@ -2435,3 +2435,18 @@ def set_category(max_sus_wind, wind_unit='kn', saffir_scale=None):
         return (np.argwhere(max_wind < saffir_scale) - 1)[0][0]
     except IndexError:
         return -1
+
+def _zlib_from_DataArray(data_var: xr.DataArray) -> bool:
```
Comment (peanutfun) [inline climada/hazard/tc_tracks.py:2439]: Linter dislikes uppercase in function names
```suggestion
def _zlib_from_dataarray(data_var: xr.DataArray) -> bool:
```

File: climada/hazard/test/test_trop_cyclone.py [hunk 4]
```diff
@@ -26,25 +26,21 @@
 
 import numpy as np
 from scipy import sparse
-import xarray as xr
 
+from climada.hazard import test as hazard_test
```
Comment (peanutfun) [inline climada/hazard/test/test_trop_cyclone.py:30]: Was there a reason to change this? I find the old version clearer:
```suggestion
import climada.hazard.test as hazard_test
```

File: climada/hazard/trop_cyclone/trop_cyclone_windfields.py [hunk 5]
```diff
@@ -1,53 +1,22 @@
...
-
-Define TC wind hazard (TropCyclone class).
-"""
-
-__all__ = ['TropCyclone']
-
-import copy
-import datetime as dt
-import itertools
 import logging
```
Comment (peanutfun) [inline climada/hazard/trop_cyclone/trop_cyclone_windfields.py:1]: Please add a module docstring to this file (or rather: revert the deletion of it)

File: climada/hazard/trop_cyclone/trop_cyclone.py [hunk 6]
```diff
@@ -0,0 +1,696 @@
...
+from tqdm import tqdm
+import pathos.pools
+import xarray as xr
+
+from climada.hazard.base import Hazard
+from climada.hazard.tc_tracks import TCTracks
+from climada.hazard.tc_clim_change import get_knutson_criterion, calc_scale_knutson
+from climada.hazard.centroids.centr import Centroids
+from climada.hazard.trop_cyclone.trop_cyclone_windfields import DEF_MAX_DIST_EYE_KM, DEF_INTENSITY_THRES, \
+    DEF_MAX_MEMORY_GB, compute_windfields_sparse
```
Comment (peanutfun) [inline climada/hazard/trop_cyclone/trop_cyclone.py:43]: line too long, why not use local import?
```suggestion
from .trop_cyclone_windfields import DEF_MAX_DIST_EYE_KM, DEF_INTENSITY_THRES, \
    DEF_MAX_MEMORY_GB, compute_windfields_sparse
```

File: climada/hazard/trop_cyclone/trop_cyclone.py [hunk 7]
```diff
@@ -0,0 +1,696 @@
...
+from climada.util import ureg
```
Comment (peanutfun) [inline climada/hazard/trop_cyclone/trop_cyclone.py:44]: unused import
```suggestion
```

File: climada/hazard/trop_cyclone/__init__.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+from climada.hazard.trop_cyclone.trop_cyclone import *
+from climada.hazard.trop_cyclone.trop_cyclone_windfields import compute_windfields_sparse, compute_angular_windspeeds, tctrack_to_si, \
+    get_close_centroids, KN_TO_MS, KM_TO_M, KM_TO_M, H_TO_S, NM_TO_KM, KMH_TO_MS, MBAR_TO_PA, \
+    DEF_MAX_DIST_EYE_KM, DEF_INTENSITY_THRES, DEF_MAX_MEMORY_GB, MODEL_VANG, DEF_RHO_AIR, DEF_GRADIENT_TO_SURFACE_WINDS, \
+    T_ICE_K, V_ANG_EARTH
```
Comment (peanutfun) [inline climada/hazard/trop_cyclone/__init__.py:5]: Does it make sense to "expose" all of these constants here? Are they, or can they be, used in other contexts at all? 🤔

File: climada/hazard/trop_cyclone/__init__.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+from climada.hazard.trop_cyclone.trop_cyclone import *
+from climada.hazard.trop_cyclone.trop_cyclone_windfields import compute_windfields_sparse, compute_angular_windspeeds, tctrack_to_si, \
```
Comment (peanutfun) [inline climada/hazard/trop_cyclone/__init__.py:2]: Use local imports
```suggestion
from .trop_cyclone import *
from .trop_cyclone_windfields import compute_windfields_sparse, compute_angular_windspeeds, tctrack_to_si, \
```

**Review by peanutfun — CHANGES_REQUESTED** [review]
Very good work! It is now much easier to get an overview of the module. I have suggestions that should fix some linter issues.

The change in module structure broke the [module documentation](https://climada-python--911.org.readthedocs.build/en/911/climada/climada.hazard.html#module-climada.hazard.trop_cyclone). You will have to add two `automodule` directives for the new submodules in the last section of `doc/climada/climada.hazard.rst`. If you are confused say so, and I will try to fix it before merging. Thanks! 🙏

File: climada/hazard/test/test_trop_cyclone.py (hunk 4 above)
Comment (ThomasRoosli) [reply climada/hazard/test/test_trop_cyclone.py]: No, this was unintentional.

File: climada/hazard/trop_cyclone/__init__.py (hunk 9 above)
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone/__init__.py:2]: fine by me (I assume this is the pattern for init files)

File: climada/hazard/trop_cyclone/__init__.py (hunk 8 above)
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone/__init__.py:5]: The constants are not private and were exposed before using the * import in the init file of hazard. So we basically allowed this usage before, and it happened e.g. in petals: https://github.com/CLIMADA-project/climada_petals/blob/main/climada_petals/hazard/tc_tracks_forecast.py#L50. I just expose them the same as before (except for the private members, which I no longer expose now).

Do you have another suggestion?

File: climada/hazard/trop_cyclone/trop_cyclone_windfields.py (hunk 5 above)
Comment (ThomasRoosli) [reply climada/hazard/trop_cyclone/trop_cyclone_windfields.py:1]: This was unintenionally not added to the newly created second file. I will add it now

**Review by ThomasRoosli — COMMENTED** [review]
Thanks for your suggestions. I agree with them. I am happy to further discuss the exposed constants.

File: climada/hazard/trop_cyclone/__init__.py (hunk 8 above)
Comment (peanutfun) [reply climada/hazard/trop_cyclone/__init__.py:5]: Alright, that's fine! I was simply not aware of this use case. Let's keep exposing them for backwards compatibility.

**Review by peanutfun — APPROVED** [review]

**Pull request merged**